src/middlewared/middlewared/client/client.py

```diff
@@ -48,126 +44,45 @@ def wait(self, timeout=None):
-                    self.reserved_fd = self.get_reserved_portfd(blacklist=reserved_ports_blacklist)
-                    sock = socket.fromfd(self.reserved_fd, family, socktype, proto)
-                    sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
-                    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-                    if hasattr(socket, 'AF_INET6') and family == socket.AF_INET6 and self.host.startswith('::'):
-                        try:
-                            sock.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_V6ONLY, 0)
-                        except (AttributeError, socket.error):
-                            pass
-                except Exception as e:
-                    if self.reserved_fd:
-                        try:
-                            os.close(self.reserved_fd)
-                        except OSError:
-                            pass
-                    raise e
-
-            WebSocket.__init__(self, sock, protocols=kwargs.get('protocols'),
-                               extensions=kwargs.get('extensions'),
-                               heartbeat_freq=kwargs.get('heartbeat_freq'))
-
-            self.stream.always_mask = True
-            self.stream.expect_masking = False
-            self.key = b64encode(os.urandom(16))
-            self._th = threading.Thread(target=self.run, name='WebSocketClient')
-            self._th.daemon = True
-
-    def get_reserved_portfd(self, blacklist=None):
+        super(WSClient, self).__init__(url, *args, **kwargs)
+
+    def bind_to_privileged_port(self):
         """
-        Get a file descriptor with a reserved port (<=1024).
-        The port is arbitrary using the libc "rresvport" call and its tried
-        again if its within `blacklist` list.
+        We always blacklist port 860 because that's the port used by iSCSI ALUA
+        on TrueNAS HA systems.
         """
-        if blacklist is None:
-            blacklist = []
-
-        libc = ctypes.cdll.LoadLibrary('libc.so.7')
-        port = ctypes.c_int(0)
-        pport = ctypes.pointer(port)
-        fd = libc.rresvport(pport)
-        retries = 5
-        while True:
-            if retries == 0:
-                break
-            if fd < 0:
-                time.sleep(0.1)
-                fd = libc.rresvport(pport)
-                retries -= 1
-                continue
-            if pport.contents.value in blacklist:
-                oldfd = fd
-                fd = libc.rresvport(pport)
-                os.close(oldfd)
-                retries -= 1
+        if self.privileged_ports_blacklist is None:
+            self.privileged_ports_blacklist = [860]
+        else:
+            self.privileged_ports_blacklist = list(set(self.privileged_ports_blacklist + [860]))
+
+        for port in (x for x in range(600, 1024) if x not in self.privileged_ports_blacklist):
+            try:
+                self.sock.bind(('', port))
+            except Exception:
```

Comment:
Exception seems too generic here, what kind of expected exceptions will it raise?

Reply:
Also, does this by default try to reuse ports? (e.g. tcp connections that are closed but still on timeout state)

Reply:
> Also, does this by default try to reuse ports? (e.g. tcp connections that are closed but still on timeout state)

Great question. In summary, I have not changed what `ws4py` is doing. However, I do believe there is a bug in the `Client\__init__` code in that library. I'll try and explain.

There are 2 scenarios.

1. The standby controller is offline.
--In this scenario, all new connections will try to bind using a local port however the TCP connections never get past the `SYN_SENT` portion of the handshake. So the `TIME_WAIT` state is never reached.

2. The standby controller is online and connections are established and then suddenly goes offline.
--In this scenario, the existing connections will go into `TIMED_WAIT` because the server did not gracefully close the existing connections.
--All new local client connections can and will use the SAME local port, however, it will not work to the same remote connection.
(i.e. bound to 169.254.10.1:600 -> 169.254.10.2:6000 -> TIMED_WAIT
       bound to 169.254.10.1:600 -> 169.254.10.2:6000 -> WILL NOT WORK because it's to the same destination)

Essentially, I've kept the same logic as what was there. However, I do believe that specifying `SO_REUSEADDR` on client based connections is incorrect and is a bug with ws4py. (which can be fixed as well.)

Reply:
There is a 3rd situation that is a little more subtle, that might even explain some other issues.

3. The standby controllers middlewared process restarts or crashes and then restarts BEFORE the `TIMED_WAIT` sockets on the client side close.
--In this scenario, any buffered data in the `TIMED_WAIT` state will push that data to the server and if the client has more than 1 local socket that has buffered data, how does the server know which data to act on in order?